### PR TITLE
podman pod create --pid flag

### PR DESCRIPTION
**Pull request by cdoern**
added support for --pid flag. User can specify ns:file, pod, private, or host pid namespace values. container returns an error since you cannot point the ns of the pods infra container
to a container outside of the pod.

Signed-off-by: cdoern <cdoern@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
What does "pod" mean?  Are you allowing pod's to join other pods pid namespace?

File: docs/source/markdown/podman-pod-create.1.md [hunk 1]
```diff
@@ -120,6 +120,15 @@ Add a DNS alias for the container. When the container is joined to a CNI network
...
 Disable creation of /etc/hosts for the pod.
 
+#### **--pid**=*pid*
+
+Set the PID mode for the pod Default is to create a private PID namespace for the pod
+
+    host: use the host’s PID namespace for the pod
+    ns: join the specified PID namespace
+    private: create a new namespace for the pod (default)
+    pod: The pod's PID namespace will be used for all containers
```
Comment (mheon) [inline docs/source/markdown/podman-pod-create.1.md:130]: I would just omit this... We should still accept it, but it doesn't need to but documented

File: docs/source/markdown/podman-pod-create.1.md [hunk 2]
```diff
@@ -120,6 +120,15 @@ Add a DNS alias for the container. When the container is joined to a CNI network
 
 Disable creation of /etc/hosts for the pod.
 
+#### **--pid**=*pid*
+
+Set the PID mode for the pod Default is to create a private PID namespace for the pod
```
Comment (mheon) [inline docs/source/markdown/podman-pod-create.1.md:125]: Should mention this requires the PID namespace be shared

File: libpod/define/pod_inspect.go [hunk 3]
```diff
@@ -57,6 +57,8 @@ type InspectPodData struct {
 	CPUQuota int64 `json:"cpu_quota,omitempty"`
 	// CPUSetCPUs contains linux specific CPU data for the pod
 	CPUSetCPUs string `json:"cpuset_cpus,omitempty"`
+	// Pid contains the process id NS for the pod
```
Comment (mheon) [inline libpod/define/pod_inspect.go:60]: "Pid is the PID namespace mode of the pod's infra container"

File: libpod/define/pod_inspect.go [hunk 4]
```diff
@@ -103,6 +105,8 @@ type InspectPodInfraConfig struct {
 	CPUQuota int64 `json:"cpu_quota,omitempty"`
 	// CPUSetCPUs contains linux specific CPU data for the container
 	CPUSetCPUs string `json:"cpuset_cpus,omitempty"`
+	// Pid contains the process id namespace information for the container
+	Pid string `json:"pid,omitempty"`
```
Comment (mheon) [inline libpod/define/pod_inspect.go:109]: This only needs to be in one place - I think here is probably the more sensible one

File: libpod/options.go [hunk 5]
```diff
@@ -2397,3 +2398,14 @@ func WithPodCPUSetCPUs(inp string) PodCreateOption {
 		return nil
 	}
 }
+
+func WithPid(inp specgen.Namespace) PodCreateOption {
```
Comment (mheon) [inline libpod/options.go:2402]: Rename to `WithPodPidNS(...)`

File: libpod/options.go [hunk 6]
```diff
@@ -2397,3 +2398,14 @@ func WithPodCPUSetCPUs(inp string) PodCreateOption {
...
+	return func(pod *Pod) error {
+		if pod.valid {
+			return define.ErrPodFinalized
+		}
+		pod.config.Pid = inp
```
Comment (mheon) [inline libpod/options.go:2407]: We need to require that `UsePodPID` is set in the pod configuration - we have to be sharing the pod PID namespace for this to be set.

File: libpod/pod.go [hunk 7]
```diff
@@ -72,6 +73,8 @@ type PodConfig struct {
 
 	// ID of the pod's lock
 	LockID uint32 `json:"lockID"`
+	// Pid contains the pod's process id namespace
+	Pid specgen.Namespace `json:"pid,omitempty"`
```
Comment (mheon) [inline libpod/pod.go:77]: This should only be in the infra container configuration, no need to duplicate

File: libpod/pod.go [hunk 8]
```diff
@@ -170,6 +174,16 @@ func (p *Pod) CPUQuota() int64 {
 	return 0
 }
 
+// Pid returns the pod process id namespace value
```
Comment (mheon) [inline libpod/pod.go:177]: This isn't getting the PID, but rather the path of the infra container's PID file?

It's also not really related to the PID namespace

File: libpod/pod_api.go [hunk 9]
```diff
@@ -541,6 +541,9 @@ func (p *Pod) Inspect() (*define.InspectPodData, error) {
 		infraConfig.CPUPeriod = p.CPUPeriod()
 		infraConfig.CPUQuota = p.CPUQuota()
 		infraConfig.CPUSetCPUs = p.ResourceLim().CPU.Cpus
+		if p.config.Pid.NSMode != "" {
+			infraConfig.Pid = p.Pid()
```
Comment (mheon) [inline libpod/pod_api.go:545]: Why is this needed?

File: libpod/runtime_pod_infra_linux.go [hunk 10]
```diff
@@ -145,6 +146,34 @@ func (r *Runtime) makeInfraContainer(ctx context.Context, p *Pod, imgName, rawIm
 		if len(p.config.InfraContainer.ExitCommand) > 0 {
 			options = append(options, WithExitCommand(p.config.InfraContainer.ExitCommand))
 		}
+		if p.config.Pid.NSMode != "private" && p.config.Pid.NSMode != "" {
+			path := p.config.Pid.Value
+			ns := spec.LinuxNamespace{}
+			pidVal := p.config.Pid.NSMode
+			if pidVal != "" {
+				if pidVal == "container" {
```
Comment (mheon) [inline libpod/runtime_pod_infra_linux.go:154]: This validation should be done in the option (`WithPid()` so we only have to do it once and we can reject invalid configurations at creation time, not runtime.

File: libpod/runtime_pod_infra_linux.go [hunk 11]
```diff
@@ -145,6 +146,34 @@ func (r *Runtime) makeInfraContainer(ctx context.Context, p *Pod, imgName, rawIm
...
+					err := errors.New("Invalid Argument")
+					return nil, errors.Wrap(err, "Cannot take container in a different NS as an argument")
+				}
+				if pidVal == "ns" {
```
Comment (mheon) [inline libpod/runtime_pod_infra_linux.go:158]: What is this if statement doing? The PID namespace and PID file are not related.

File: libpod/runtime_pod_infra_linux.go [hunk 12]
```diff
@@ -145,6 +146,34 @@ func (r *Runtime) makeInfraContainer(ctx context.Context, p *Pod, imgName, rawIm
...
+					path = p.config.Pid.Value
+					_, err := os.Stat("path")
+					if err != nil {
+						return nil, err
+					}
+					options = append(options, WithPidFile(path))
+				}
+			}
+			if pidVal != "host" {
```
Comment (mheon) [inline libpod/runtime_pod_infra_linux.go:167]: I think you probably want to structure this as a `switch` with a blog for each valid option

File: libpod/runtime_pod_infra_linux.go [hunk 13]
```diff
@@ -145,6 +146,34 @@ func (r *Runtime) makeInfraContainer(ctx context.Context, p *Pod, imgName, rawIm
...
+				ns.Type = "pid"
+				ns.Path = path
+				g.AddOrReplaceLinuxNamespace(string(ns.Type), ns.Path)
+			} else if pidVal == "host" {
+				g.Config.Linux.Namespaces = []spec.LinuxNamespace{}
```
Comment (mheon) [inline libpod/runtime_pod_infra_linux.go:172]: This wipes out *all* namespaces, not just the PID namespace. You need to loop through namespaces and make a new array containing everything except the PID namespace.

File: libpod/runtime_pod_linux.go [hunk 14]
```diff
@@ -116,6 +116,10 @@ func (r *Runtime) NewPod(ctx context.Context, options ...PodCreateOption) (_ *Po
 	if pod.config.UsePodCgroup {
 		logrus.Debugf("Got pod cgroup as %s", pod.state.CgroupPath)
 	}
+
+	//	if !pod.state.Pid.IsDefault(){
```
Comment (mheon) [inline libpod/runtime_pod_linux.go:120]: This looks like it can be removed

File: pkg/domain/entities/pods.go [hunk 15]
```diff
@@ -146,6 +147,17 @@ func (p *PodCreateOptions) CPULimits() *specs.LinuxCPU {
 	return cpu
 }
 
+func setNamespaces(p *PodCreateOptions) (specgen.Namespace, error) {
```
Comment (mheon) [inline pkg/domain/entities/pods.go:150]: Suggest renaming to "getPIDNamespace" or, alternatively, making this return a struct which identifies which namespace is which - right now this is only useful for setting the PID namespace, but eventually we'll end up with four-ish settable namespaces.

File: pkg/specgen/generate/pod_create.go [hunk 16]
```diff
@@ -102,6 +102,13 @@ func createPodOptions(p *specgen.PodSpecGenerator, rt *libpod.Runtime) ([]libpod
 		options = append(options, libpod.WithInfraCommand(p.InfraCommand))
 	}
 
+	if !p.Pid.IsDefault() {
+		options = append(options, libpod.WithPid(p.Pid))
+		if p.Pid.NSMode == "pod" {
```
Comment (mheon) [inline pkg/specgen/generate/pod_create.go:107]: This bit doesn't seem necessary?

File: pkg/domain/entities/pods.go [hunk 17]
```diff
@@ -178,6 +190,14 @@ func (p *PodCreateOptions) ToPodSpecGen(s *specgen.PodSpecGenerator) error {
 	s.NoManageHosts = p.Net.NoHosts
 	s.HostAdd = p.Net.AddHosts
 
+	pidR, err := setNamespaces(p)
```
Comment (mheon) [inline pkg/domain/entities/pods.go:193]: Minor nit, but slightly more descriptive variable names would be nice - I don't really like having a function with `pid` and `pidR`

File: pkg/specgen/podspecgen.go [hunk 18]
```diff
@@ -57,6 +57,8 @@ type PodBasicConfig struct {
 	// (e.g. `podman generate systemd --new`).
 	// Optional.
 	PodCreateCommand []string `json:"pod_create_command,omitempty"`
+	// Pid sets the process id namespace of the pod
```
Comment (mheon) [inline pkg/specgen/podspecgen.go:60]: Add a line underneath - "Optional (defaults to private if unset)". And this should mention that it sets the PID namespace of the infra container, that will then be shared with the entire pod if PID namespace sharing is enabled.

File: libpod/options.go (hunk 6 above)
Comment (cdoern) [reply libpod/options.go:2407]: > We need to require that `UsePodPID` is set in the pod configuration - we have to be sharing the pod PID namespace for this to be set.

Should I require it by checking and erroring out if not set, or warn the user and set it for them?

File: libpod/pod.go (hunk 8 above)
Comment (cdoern) [reply libpod/pod.go:177]: Yeah I was unsure here of what would be helpful here @mheon, because the namespace just contains `pod` or `host` or whatever the user passed or a ns file if specified. What would this be expected to return?

File: libpod/pod_api.go (hunk 9 above)
Comment (cdoern) [reply libpod/pod_api.go:545]: If I can fix `Pid()` to return the proper value I guess it won't be but right now it is never "empty" so the file path is always in inspect even when user didn't specify `--pid`

File: libpod/options.go (hunk 6 above)
Comment (mheon) [reply libpod/options.go:2407]: Good question. I was leaning towards just erroring out unless the user passes `--share` as well to specify that the PID namespace is shared, but that seems like a bad user experience. @rhatdan Should setting `--pid` automatically cause the pod to share the PID namespace?

File: libpod/pod.go (hunk 8 above)
Comment (mheon) [reply libpod/pod.go:177]: Returning the mode of the namespace is what I would expect - we want to tell the user how the pod was configured. The PID of the pod's init container isn't that valuable.

File: libpod/runtime_pod_infra_linux.go (hunk 10 above)
Comment (cdoern) [reply libpod/runtime_pod_infra_linux.go:154]: Okay @mheon , to streamline this and preserve the originally passed values I might make another InfraContainerConfig value that is an actual `specs.LinuxNameSpace{}`. This will allow me to be able to print the proper mode in podInspect but also not have to parse anything again in `runtime_pod_infra_linux.go`

File: libpod/options.go [hunk 19]
```diff
@@ -2397,3 +2400,32 @@ func WithPodCPUSetCPUs(inp string) PodCreateOption {
...
+				err := errors.New("Invalid Argument")
+				return errors.Wrap(err, "Cannot take container in a different NS as an argument")
+			case "path":
+				fmt.Println("hello")
+				ns.Path = inp.Value
+			case "private":
+				ns.Type = spec.PIDNamespace
+				ns.Path = ""
+			}
+			fmt.Println(ns)
```
Comment (mheon) [inline libpod/options.go:2424]: Extra debug

File: libpod/options.go [hunk 20]
```diff
@@ -2397,3 +2400,32 @@ func WithPodCPUSetCPUs(inp string) PodCreateOption {
...
 }
+
+func WithPodPidNS(inp specgen.Namespace) PodCreateOption {
+	return func(p *Pod) error {
+		if p.valid {
+			return define.ErrPodFinalized
+		}
+		if p.config.UsePodPID {
+			ns := specs.LinuxNamespace{}
+			ns.Type = "pid"
```
Comment (mheon) [inline libpod/options.go:2411]: Is there a reason we need to recreate the namespace?

File: libpod/options.go [hunk 21]
```diff
@@ -2397,3 +2400,32 @@ func WithPodCPUSetCPUs(inp string) PodCreateOption {
...
+			fmt.Println(inp)
+			switch inp.NSMode {
+			case "container":
+				err := errors.New("Invalid Argument")
```
Comment (mheon) [inline libpod/options.go:2415]: use define.ErrInvalidArg instead

File: libpod/options.go (hunk 20 above)
Comment (cdoern) [reply libpod/options.go:2410]: I think `specs.LinuxNameSpace{}` is the type that the infra generator takes versus `specgen.NameSpace`, we need to switch over at some point and this is behind the scenes enough.

File: libpod/options.go (hunk 19 above)
Comment (cdoern) [reply libpod/options.go:2423]: whoops!

File: libpod/options.go (hunk 20 above)
Comment (mheon) [reply libpod/options.go:2409]: Ahhh, I missed the type change. Hm. My concern would be forwards-compat - we need to make sure that pods from before this change, without a valid namespace in the DB, continue to work.

Comment (cdoern) [reply libpod/options.go:2409]: well,  the `specgen.ParseNamespace()` function I use in pkg/domain/entities/pods.go takes the type specgen.NameSpace so the actual validation and parsing relies more on the specgen version this is just a conversion. This switch will only execute when someone explicitly says `--pid=...` so would forwards-compat be a problem?

**Comment by cdoern** [comment]
> What does "pod" mean? Are you allowing pod's to join other pods pid namespace?

pod means that the containers will default to using the pod's pid ns which is similar to private (the default). I added this because the `specgen.parsenamespace()` function utilizes this value so we are going to accept.

File: libpod/pod.go [hunk 22]
```diff
@@ -97,6 +98,8 @@ type InfraContainerConfig struct {
 	HasInfraContainer  bool                  `json:"makeInfraContainer"`
 	NoNetwork          bool                  `json:"noNetwork,omitempty"`
 	HostNetwork        bool                  `json:"infraHostNetwork,omitempty"`
+	Pid                specgen.Namespace     `json:"infraPid,omitempty"`
+	LinuxPidNS         specs.LinuxNamespace  `json:"LinuxPIDNS,omitempty"`
```
Comment (mheon) [inline libpod/pod.go:102]: Do we need both of these? I would imagine we can get the mode just by looking at the `specs.LinuxNamespace` - or, alternatively, generate the `specs.LinuxNamespace` from the `specgen.Namespace` when creating the infra container?

Comment (cdoern) [reply libpod/pod.go:102]: @mheon I kept both of them in order to parse the Pid NS before runtime. The types are different as `specgen.Namespace` contains info like `host` or `pod` which I do use again at runtime but `specs.LinuxNamespace` is the actual NS struct containing `pid` and the proper path. Not sure if I can parse before runtime and not have both of these but I will look.

Comment (cdoern) [reply libpod/pod.go]: fixed

File: docs/source/markdown/podman-pod-create.1.md [hunk 23]
```diff
@@ -120,6 +120,14 @@ Add a DNS alias for the container. When the container is joined to a CNI network
 
 Disable creation of /etc/hosts for the pod.
 
+#### **--pid**=*pid*
+
+Set the PID mode for the pod Default is to create a private PID namespace for the pod. Requires PID NS to be shared.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-pod-create.1.md:125]: I'm having a hard time parsing this.  Maybe?
```suggestion
Set the PID mode for the pod.  The default is to create a private PID namespace for the pod. Requires the PID namespace to be shared.
```

File: docs/source/markdown/podman-pod-create.1.md [hunk 24]
```diff
@@ -120,6 +120,14 @@ Add a DNS alias for the container. When the container is joined to a CNI network
 
 Disable creation of /etc/hosts for the pod.
 
+#### **--pid**=*pid*
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-pod-create.1.md:123]: I was thinking "process ID (pid) when reading this.  I'll let others over-rule me, but I'd rather the option be named something else, in case we ever manipulate the pid later down the road.  Perhaps `--pid-mode`, regardless, at least  change the variable:
```suggestion
#### **--pid**=*pid-mode*
```

File: libpod/pod.go (hunk 8 above)
Comment (TomSweeneyRedHat) [reply libpod/pod.go:175]: Similar comment here for me, I'd change the variable to indicate it's pidmode and not the pid to help make the support of this more clear months down the line.

File: docs/source/markdown/podman-pod-create.1.md (hunk 24 above)
Comment (cdoern) [reply docs/source/markdown/podman-pod-create.1.md:123]: I feel like `--pid-ns` might be more accurate if anything @TomSweeneyRedHat. Only thing is, [container create](http://docs.podman.io/en/latest/markdown/podman-create.1.html#pid-pid) calls it `--pid`

File: libpod/pod.go (hunk 8 above)
Comment (cdoern) [reply libpod/pod.go:175]: good idea @TomSweeneyRedHat. I will change to make it more explicit that it returns the mode passed by the user.

File: docs/source/markdown/podman-pod-create.1.md (hunk 24 above)
Comment (TomSweeneyRedHat) [reply docs/source/markdown/podman-pod-create.1.md:123]: Ugh.  They should be the same in both places, and it looks like Docker has gone with `--pid` too, so we probably ought to stay with `--pid`.  Rather than `*pid-mode*` at line 123, I'd be fine with `*pid-ns*` if you prefer for the variable

Comment (cdoern) [reply docs/source/markdown/podman-pod-create.1.md:123]: okay @TomSweeneyRedHat I am good with keeping it as `--pid`, probably best so people don't get confused as to what this flag is compared to container create and docker

File: libpod/pod.go [hunk 25]
```diff
@@ -97,6 +98,7 @@ type InfraContainerConfig struct {
 	HasInfraContainer  bool                  `json:"makeInfraContainer"`
 	NoNetwork          bool                  `json:"noNetwork,omitempty"`
 	HostNetwork        bool                  `json:"infraHostNetwork,omitempty"`
+	Pid                specgen.Namespace     `json:"infraPid,omitempty"`
```
Comment (vrothberg) [inline libpod/pod.go:101]: I think this should be called `PidNS` to prevent mistaking it for a process ID.  Same at other places in the code.

But I do agree that `--pid` makes sense on the CLI to remain consistent with `container create`.

File: test/e2e/pod_create_test.go [hunk 26]
```diff
@@ -559,4 +559,37 @@ ENTRYPOINT ["sleep","99999"]
...
+		podCreate = podmanTest.Podman([]string{"pod", "create", "--pid", ns, "--share", "pid"})
+		podCreate.WaitWithDefaultTimeout()
+		Expect(podCreate.ExitCode()).To(Equal(0))
+
+		ns = "container:randomfakeid"
+
+		podCreate = podmanTest.Podman([]string{"pod", "create", "--pid", ns, "--share", "pid"})
+		podCreate.WaitWithDefaultTimeout()
+		Expect(podCreate).Should(ExitWithError())
+
```
Comment (vrothberg) [inline test/e2e/pod_create_test.go:593]: Will the pid ns now be displayed in `podman inspect`? If so, it would be nice to exercise that here in the test.

**Review by vrothberg — COMMENTED** [review]
LGTM in general. Just two comments.

File: test/e2e/pod_create_test.go (hunk 26 above)
Comment (cdoern) [reply test/e2e/pod_create_test.go:593]: @vrothberg we decided to output the `NSMode` the user gave us as it is more helpful, aka: `path`, `pod`, `private`... I will check for that here, thanks for calling that out!

**Comment by cdoern** [comment]
sorry had to re-push to fix a test, ran locally and all green. should be good now @baude
